apps/call.go

```diff
@@ -7,72 +7,110 @@ import (
 type CallType string
 
 const (
 	// CallTypeSubmit (default) indicates the intent to take action.
-	CallTypeSubmit = CallType("submit")
+	CallTypeSubmit CallType = "submit"
+
 	// CallTypeForm retrieves the form definition for the current set of values,
 	// and the context.
-	CallTypeForm = CallType("form")
+	CallTypeForm CallType = "form"
+
 	// CallTypeCancel is used for for the (rare?) case of when the form with
 	// SubmitOnCancel set is dismissed by the user.
-	CallTypeCancel = CallType("cancel")
+	CallTypeCancel CallType = "cancel"
+
 	// CallTypeLookup is used to fetch items for dynamic select elements
-	CallTypeLookup = CallType("lookup")
+	CallTypeLookup CallType = "lookup"
 )
 
-// A Call invocation is supplied a BotAccessToken as part of the context. If a
-// call needs acting user's or admin tokens, it should be specified in the
-// Expand section.
-//
-// If a user or admin token are required and are not available from previous
-// consent, the appropriate OAuth flow is launched, and the Call is executed
-// upon its success.
-//
-// TODO: what if a call needs a token and it was not provided? Return a call to
-// itself with Expand.
+// Call defines a way to invoke an App's function. Calls are used to fetch App's
+// bindings, to process notifications, and to respond to user input from forms,
+// bindings and command line.
 type Call struct {
-	Path   string      `json:"path,omitempty"`
-	Expand *Expand     `json:"expand,omitempty"`
-	State  interface{} `json:"state,omitempty"`
+	// The path of the Call. For HTTP apps, the path is appended to the app's
+	// RootURL. For AWS Lambda apps, it is mapped to the appropriate Lambda name
+	// to invoke, and then passed in the call request.
+	Path string `json:"path,omitempty"`
+
+	// Expand specifies what extended data should be provided to the function in
+	// each request's Context. It may be various auth tokens, configuration
+	// data, or details of Mattermost entities such as the acting user, current
+	// team and channel, etc.
+	Expand *Expand `json:"expand,omitempty"`
+
+	// Custom data that will be passed to the function in JSON, "as is".
+	State interface{} `json:"state,omitempty"`
 }
 
+// CallRequest envelops all requests sent to Apps.
 type CallRequest struct {
+	// A copy of the Call struct that originated the request. Path and State are
+	// of significance.
 	Call
-	Type          CallType               `json:"type"`
-	Values        map[string]interface{} `json:"values,omitempty"`
-	Context       *Context               `json:"context,omitempty"`
-	RawCommand    string                 `json:"raw_command,omitempty"`
-	SelectedField string                 `json:"selected_field,omitempty"`
-	Query         string                 `json:"query,omitempty"`
+
+	// Type of the request, see CallType type for more information.
+	Type CallType `json:"type"`
+
+	// Values are all values entered by the user.
+	Values map[string]interface{} `json:"values,omitempty"`
```

Comment:
Could we envision a case in which they are not "entered by the user"?

Reply:
If we add some kind of "hidden field" that could be information not entered by the user.
In theory, there is no need to add a hidden field, since we have the state, but maybe it makes sense.